crates/typst/src/layout/grid/lines.rs

```diff
@@ -0,0 +1,749 @@
+/// stroke (with priority to the vline's stroke, followed by the right cell's
+/// stroke, and, finally, the left cell's) and returned. If, however, the cells
+/// around the vline at this row do not have any stroke overrides, then the
+/// vline's own stroke is directly returned.
+pub(super) fn vline_stroke_at_row(
+    grid: &CellGrid,
+    x: usize,
+    y: usize,
+    stroke: Option<Arc<Stroke<Abs>>>,
+) -> Option<Arc<Stroke<Abs>>> {
+    if x != 0 && x != grid.cols.len() {
+        // When the vline isn't at the border, we need to check if a colspan would
+        // be present between columns 'x' and 'x-1' at row 'y', and thus overlap
+        // with the line.
+        // To do so, we analyze the cell right after this vline. If it is merged
+        // with a cell before this line (parent_x < x) which is at this row or
+        // above it (parent_y <= y), this means it would overlap with the vline,
+        // so the vline must not be drawn at this row.
+        let first_adjacent_cell = if grid.has_gutter {
+            // Skip the gutters, if x or y represent gutter tracks.
+            // We would then analyze the cell one column after (if at a gutter
+            // column), and/or one row below (if at a gutter row), in order to
+            // check if it would be merged with a cell before the vline.
+            (x + x % 2, y + y % 2)
+        } else {
+            (x, y)
+        };
+        let Axes { x: parent_x, y: parent_y } = grid
+            .parent_cell_position(first_adjacent_cell.0, first_adjacent_cell.1)
+            .unwrap();
+
+        if parent_x < x && parent_y <= y {
+            // There is a colspan cell going through this vline's position,
+            // so don't draw it here.
+            return None;
+        }
+    }
+
+    let left_cell_stroke = x
+        .checked_sub(1)
+        .and_then(|left_x| grid.parent_cell(left_x, y))
+        .and_then(|left_cell| left_cell.stroke.right.as_ref());
+    let right_cell_stroke = if x < grid.cols.len() {
+        grid.parent_cell(x, y)
+            .and_then(|right_cell| right_cell.stroke.left.as_ref())
+    } else {
+        None
+    };
+
+    let cell_stroke = match (left_cell_stroke.cloned(), right_cell_stroke.cloned()) {
+        (Some(left_cell_stroke), Some(right_cell_stroke)) => {
+            // When both cells specify a stroke for this line segment, fold
+            // both strokes, with priority to the right cell's left stroke.
+            Some(right_cell_stroke.fold(left_cell_stroke))
+        }
+        // When one of the cells doesn't specify a stroke, the other cell's
+        // stroke should be used.
+        (left_cell_stroke, right_cell_stroke) => left_cell_stroke.or(right_cell_stroke),
+    };
+
+    // Fold the line stroke and folded cell strokes, if possible.
+    // Give priority to the explicit line stroke.
+    // Otherwise, use whichever of the two isn't 'none' or unspecified.
+    match (cell_stroke, stroke) {
```

Comment:
This here seems to be the exact same thing as above and in the function below, we have it twice again. Maybe we should have a function for it?

It also comes up in other places I think. It's sort of an "alternative" Option fold.

Reply:
Yeah I was thinking that too. I'll probably make a function for it.

Reply:
I've created the `AlternativeFold` trait, which provides `fold_or` for `Option<T>` when `T` is foldable.